### PR TITLE
[ctx_prof] Fix checks in `PGOCtxprofFlattening`

### PR DESCRIPTION
**Pull request by mtrofin**
The assertion that all out-edges of a BB can't be 0 is incorrect: they can be, if that branch is on a cold subgraph.

Added validators and asserts about the expected proprerties of the propagated counters.

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 1]
```diff
@@ -198,6 +204,50 @@ class ProfileAnnotator final {
...
+    for (const auto &BBInfo : BBInfos)
+      if (!BBInfo.second.hasCount())
+        return false;
+    for (const auto &EdgeInfo : EdgeInfos)
+      if (!EdgeInfo.Count.has_value())
+        return false;
+    return true;
+  }
+
+  bool allTakenPathsExit() const {
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:223]: add a description for the function. Also make it static?

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 2]
```diff
@@ -198,6 +204,50 @@ class ProfileAnnotator final {
 
   BBInfo &getBBInfo(const BasicBlock &BB) { return BBInfos.find(&BB)->second; }
 
+  const BBInfo &getBBInfo(const BasicBlock &BB) const {
+    return BBInfos.find(&BB)->second;
+  }
+
+  // validation function after we propagate the counters: all BBs and edges'
+  // counters must have a value.
+  bool allCountersAreAssinged() const {
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:213]: make it static?

File: llvm/test/Analysis/CtxProfAnalysis/flatten-zero-path.ll [hunk 3]
```diff
@@ -0,0 +1,54 @@
+; RUN: split-file %s %t
+; RUN: llvm-ctxprof-util fromJSON --input=%t/profile.json --output=%t/profile.ctxprofdata
```
Comment (david-xl) [inline llvm/test/Analysis/CtxProfAnalysis/flatten-zero-path.ll:2]: describe the purpose of the test?

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 2 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:213]: why, it uses private members of `ProfileAnnotator`.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 1 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:223]: doc done.

re. static, same as above - it uses private members of this class.

File: llvm/test/Analysis/CtxProfAnalysis/flatten-zero-path.ll (hunk 3 above)
Comment (mtrofin) [reply llvm/test/Analysis/CtxProfAnalysis/flatten-zero-path.ll:2]: done.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 4]
```diff
@@ -58,22 +60,28 @@ class ProfileAnnotator final {
...
-      uint64_t Sum = 0;
+    // Return std::nullopt if there were no edges to sum. The user can decide
+    // how to interpret that.
+    std::optional<uint64_t> getEdgeSum(const SmallVector<EdgeInfo *> &Edges,
+                                       bool AssumeAllKnown) const {
+      std::optional<uint64_t> Sum;
       for (const auto *E : Edges)
-        if (E)
-          Sum += AssumeAllKnown ? *E->Count : E->Count.value_or(0U);
+        if (E) {
```
Comment (teresajohnson) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:69]: What is the situation where we have an entry in the Edges vector but it is null?

Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:69]: Edges that are removed, e.g. the faux coroutine suspend edge. I'll add a comment on the Edges member to explain the contents.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 5]
```diff
@@ -178,7 +191,7 @@ class ProfileAnnotator final {
     bool KeepGoing = true;
     while (KeepGoing) {
       KeepGoing = false;
-      for (const auto &BB : reverse(F)) {
+      for (const auto &BB : F) {
```
Comment (teresajohnson) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:194]: why this change?

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 6]
```diff
@@ -198,6 +211,52 @@ class ProfileAnnotator final {
 
   BBInfo &getBBInfo(const BasicBlock &BB) { return BBInfos.find(&BB)->second; }
 
+  const BBInfo &getBBInfo(const BasicBlock &BB) const {
+    return BBInfos.find(&BB)->second;
+  }
+
+  // validation function after we propagate the counters: all BBs and edges'
+  // counters must have a value.
+  bool allCountersAreAssinged() const {
```
Comment (teresajohnson) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:220]: typo s/Assinged/Assigned/

File: llvm/test/ThinLTO/X86/ctxprof.ll [hunk 7]
```diff
@@ -47,18 +47,21 @@
 ; NOPROF-1-NOT: m2_f1()
 ; NOPROF-2-NOT: m1_f1()
 ;
-; The run with workload definitions - same other options.
+; The run with workload definitions - same other options. We do need to re-generate the .bc
+; files, to include instrumentation.
```
Comment (teresajohnson) [inline llvm/test/ThinLTO/X86/ctxprof.ll:51]: How did this work before this change?

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 5 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:194]: We always have the entry BB's count, so it makes more sense to traverse top down.

File: llvm/test/ThinLTO/X86/ctxprof.ll (hunk 7 above)
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/ctxprof.ll:51]: These functions have no edges. Before, the validation had to do with edges, now we insist all counters are defined.

**Review by teresajohnson — APPROVED** [review]
lgtm
